doc/architecture/decisions/0002-git-commit-messages.md

```diff
@@ -0,0 +1,45 @@
+# 2. git commit messages
+
+Date: 2019-11-13
+
+## Status
+
+Draft
+
+## Context
+
+We need to improve the quality of the git commit messages so developers can easily understand the context of the proposed changes, throughout a concise and consistent message in order to avoid lengthy code reviews and project long term maintainability.
+
+Also, a good and structured message will support the automation of the releases notes or at least reduce the burn of it.
+
+## Decision
+
+The first interation will rely on generic rules described [here](https://chris.beams.io/posts/git-commit/), along with some specificities (e.g. GitHub issue if associated with one).
+
+Most wanted rules:
+
+1. Separate subject from body with a blank line
+2. Limit the subject line to 50 characters
+3. Capitalize the subject line
+4. Wrap the body at 72 characters
+5. Use the body to explain what and why vs. how
+
+Other rules are also important!
+
+In order to automate the release notes/changelog process, we also need to define basic semantics:
+
+| Commit message | Release type |
+|----------------|--------------|
+|fix: Stop the O2 leak dock comportment (DC)|Patch release|
+|feat: Implements dog space travel module (BARK)|Feature release|
+|break: Removes the compase from the control room (DOOM)|Breaking change release|
```

Comment:
Having a fixed prefix (here, up to 7 characters including the space) plus a maximum of 50 characters total in the subject (= 43 characters) can make expressing the intent difficult.

Looking at the actual commits though (`git log master --no-merges --pretty=%s`), our peaks of commit summaries are at 31 / 39 characters (19 and 18 commits, respectively), so that's probably okay?  There's still a pretty big tail of >43 character commit messages (177 commits out of 484), including one mistake that had 170 characters (because the new line after the subject was missing)…

Of course, that 177 out of 484 is suspicious too, because our shortest summary was `typo`…

Reply:
Thanks for the numbers...so we should be safe?! 🏝

Regarding the prefix, it would help the automation of the release notes and even the manual one when needed.

Reply:
Yep, I can see that it's useful for machines.  But machines don't need it in the _summary_; they can grep the message body just as well.  Of course, that makes it harder for _people_, so we'd need to figure out the audience for the tagging first.

Reply:
I agree that probably want to keep as much of the summary to convey purpose from human to human and could use the trailer of a commit for anything a machine should handle, similar to `Signed-off-by`.

[Git notes](https://stackoverflow.com/questions/2683248/can-i-add-metadata-to-git-commits-or-can-i-hide-some-tags-in-gitk] are probably too cumbersome to use.

Reply:
@mook-as the audience is everyone, including release managers and developers. I also tend to think that the summary should be short, clear and concise to give a clear idea to the "reader" about the intention of the commit and a prefix of 3/4 chars is ok.

For the time being and thanks to your numbers 😉, I think we should move forward unless there's a better feasible option in the meantime.

Reply:
If the target audience is _everyone_, then I'd argue that excess cruft should be as a trail instead of part of the summary line.  I'm unlikely to care when reading `git log` if I'm dealing with a bug report or a feature; I just care about what the commit intended.  I sometimes have enough trouble fitting my summary in the existing limit (50, according to `/usr/share/vim/vim80/syntax/gitcommit.vim`) already.

That is, the commit message would look like:

```
Frob the doodad.

We need to frob the doodad so that the thingy can interact with the 
widget when required.  [jsc#CAP-1234]

Type: feature
```

Now that I think about it, this (either the summary or the trailer) doesn't work well for multi-commit PRs.  We need to have explicit wording around that, I think.

Reply:
@mook-as Read my suggestion to handle multi-commit PRs.

Reply:
Squashes are _sometimes_ a good answer, but not always; occasionally it may be useful to have multiple commits if the PR is large.

… of course, we could always say that requires multiple PRs instead (one per would-be-commit); but that may not be the way to go if we need more exploratory work / some of the commits need to be amended.

Reply:
Past the alpha, big spikes should become rare.

Reply:
Well, Jan's working on the big changes to configgin still for SCF 1.5.x… so I don't think that's worked out :)